src/process/message_processor.go

```diff
@@ -1,48 +1,94 @@
-		if err != nil {
-			return errors.New("error creating parsing window, " + err.Error())
-		}
+		// Round timestamp to size of parseDuration to place in correct bucket
+		messageBucket := msg.Timestamp.Truncate(parseDuration)
+
+		// If a bucket with this timestamp doesn't exist, need to create one
+		if _, ok := window.BucketMap[messageBucket]; !ok {
+			fmt.Println("----------------------------------------------------------------------------------")
+
+			// If there's at least one timestamp, may have to fill in missing zeroes in map
+			if len(window.BucketMap) > 1 {
+				// Get minimum timestamp from map
+				minTimestamp := getBucketMapMin(window.BucketMap)
+
+				// Walk through map and see if any zeroes are missing
+				for tempTimestamp := minTimestamp; tempTimestamp.Before(messageBucket); tempTimestamp = tempTimestamp.Add(time.Minute) {
+					if _, ok := window.BucketMap[tempTimestamp]; !ok {
+						fmt.Println("Appended a 0 to map at ", tempTimestamp)
+						window.BucketMap[tempTimestamp] = make([]common.BGPMessage, 0)
+					}
+				}
+			}
+
+			// If we at least maximumBuckets in bucketMap we can perform analysis
+			if len(window.BucketMap) >= maximumBuckets {
+				fmt.Println("len(window.BucketMap):", len(window.BucketMap), " maximumBuckets:", maximumBuckets)
 
-		// Round down the timestamp to the nearest multiple of 60 seconds
-		bucketTimestamp := msg.Timestamp.Truncate(windowSize)
+				// First want to remove timestamps out of scope so len(bucketMap) == maximumBuckets
+				minimumTimestamp := messageBucket.Add(-maximumTimespan)
+				for timestamp := range window.BucketMap {
+					if timestamp.Before(minimumTimestamp) {
+						fmt.Println("Expired bucket: ", timestamp)
+						delete(window.BucketMap, timestamp)
+					}
+				}
 
-		// Check if a bucket for the rounded timestamp exists, create it if not
-		if _, ok := bucketMap[bucketTimestamp]; !ok {
-			bucketMap[bucketTimestamp] = make([]common.BGPMessage, 0)
+				// Now window is ready for analysis
+				analyze.AnalyzeBGPMessages(window)
+			}
+
+			// Create new bucket for new timestamp
+			fmt.Println("Creating bucket: ", messageBucket)
+			window.BucketMap[messageBucket] = make([]common.BGPMessage, 0)
 		}
 
 		// Append the message to the corresponding bucket
-		bucketMap[bucketTimestamp] = append(bucketMap[bucketTimestamp], msg)
+		window.BucketMap[messageBucket] = append(window.BucketMap[messageBucket], msg)
 	}
 
-	for timestamp, messages := range bucketMap {
-		frequency := len(messages)
-		fmt.Printf("Timestamp: %s, Frequency: %d\n", timestamp.Format("2006-01-02 15:04:05"), frequency)
-	}
-	fmt.Println("Minutes Analyzed: ", len(bucketMap))
+	return nil
+}
 
-	window := common.Window{
-		Filter:    "none",
-		BucketMap: bucketMap,
+// Returns the minimum key value of a bucketMap
+func getBucketMapMin(bucketMap map[time.Time][]common.BGPMessage) time.Time {
+	var minTimestamp time.Time
+	firstIteration := true
+	for timestamp := range bucketMap {
```

Comment:
I think it's more readable to initialize `minTimestamp` with the value of the first index of the map, then start from the second index in the loop (no `firstIteration` is needed)